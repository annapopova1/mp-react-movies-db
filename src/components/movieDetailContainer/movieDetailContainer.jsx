import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MoviesList from '../moviesList/moviesList';
import NavPanel from '../navPanel/navPanel';
import MovieDetail from '../movieDetail/movieDetail';
import { loadMovie, deactivateSSRFlag } from '../../store/actions/movieViewActions';

export class MovieDetailContainerUI extends Component {
  static propTypes = {
    movie: PropTypes.shape({}),
    moviesByGenre: PropTypes.arrayOf(PropTypes.shape({})),
    loadMovie: PropTypes.func,
    isMovieLoading: PropTypes.bool,
    match: PropTypes.shape({}).isRequired,
    history: PropTypes.shape({}).isRequired,
    isSSR: PropTypes.bool,
    deactivateSSRFlag: PropTypes.func,
  };

  static defaultProps = {
    movie: null,
    moviesByGenre: [],
    loadMovie: () => {},
    isMovieLoading: false,
    isSSR: true,
    deactivateSSRFlag: () => {},
  };

  componentDidMount() {
    if (this.props.isSSR) {
      this.props.deactivateSSRFlag();
    } else {
      this.props.loadMovie(this.props.match.params.movieId);
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.movieId !== prevProps.match.params.movieId) {
      this.props.loadMovie(this.props.match.params.movieId);
    }
  }

  openMovieDetail = movieId => () => {
    this.props.history.push(`/film/${movieId}`);
  };

  render() {
    const { movie, isMovieLoading, moviesByGenre } = this.props;
    return (
      <Fragment>
        {
            isMovieLoading
              ? <div className="alert alert-info">Movie is loading...</div>
              : movie
                ? <Fragment>
                  <MovieDetail movie={movie} />
                  <NavPanel
                    direction="left"
                    secondaryBrand="Films by"
                    links={[{ title: `Genre: ${movie.genres.join(', ')}`, active: true, param: 'genres' }]}
                  />
                  <MoviesList movies={moviesByGenre} openDetailHandler={this.openMovieDetail} />
                </Fragment>
                : <div className="alert alert-warning">Movie is not available :-(</div>
          }
      </Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  const { movie, isMovieLoading, moviesByGenre, isSSR } = state.movieView;
  return {
    isSSR,
    movie,
    isMovieLoading,
    moviesByGenre,
  };
};

const mapDispatchToProps = dispatch => ({
  loadMovie: (id) => {
    dispatch(loadMovie(id));
  },
  deactivateSSRFlag: () => {
    dispatch(deactivateSSRFlag());
  }
});

const MovieDetailContainer = connect(mapStateToProps, mapDispatchToProps)(MovieDetailContainerUI);
export default MovieDetailContainer;
