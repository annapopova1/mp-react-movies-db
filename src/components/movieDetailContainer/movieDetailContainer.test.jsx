import { set } from 'lodash';
import React from 'react';
import { MovieDetailContainerUI } from './movieDetailContainer';
import { MOVIES_LIST } from '../../../test/movies.test-data';

describe('<MovieDetailContainer/>', () => {
  const history = { push: jest.fn() };
  const match = {};
  set(match, 'params.movieId', '123456');

  test('should render MovieDetailContainer component', () => {
    const app = shallow(<MovieDetailContainerUI
      movie={MOVIES_LIST[0]}
      moviesByGenre={[]}
      loadMovie={() => {}}
      isMovieLoading={false}
      match={match}
      history={history}
    />);
    expect(app).toMatchSnapshot();
  });
});
