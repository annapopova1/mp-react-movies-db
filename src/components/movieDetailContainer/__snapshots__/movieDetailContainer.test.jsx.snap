// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDetailContainer/> should render MovieDetailContainer component 1`] = `
<React.Fragment>
  <React.Fragment>
    <MovieDetail
      movie={
        Movie {
          "budget": 0,
          "genres": Array [
            "genre1",
            "genre2",
          ],
          "id": 1,
          "overview": "This is a wider card with supporting text below as a natural lead-in to additional content.",
          "posterPath": "data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%22348%22%20height%3D%22225%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20348%20225%22%20preserveAspectRatio%3D%22none%22%3E%3Cdefs%3E%3Cstyle%20type%3D%22text%2Fcss%22%3E%23holder_162cee39977%20text%20%7B%20fill%3A%23eceeef%3Bfont-weight%3Abold%3Bfont-family%3AArial%2C%20Helvetica%2C%20Open%20Sans%2C%20sans-serif%2C%20monospace%3Bfont-size%3A17pt%20%7D%20%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cg%20id%3D%22holder_162cee39977%22%3E%3Crect%20width%3D%22348%22%20height%3D%22225%22%20fill%3D%22%2355595c%22%3E%3C%2Frect%3E%3Cg%3E%3Ctext%20x%3D%22116.7265625%22%20y%3D%22120.3%22%3EThumbnail%3C%2Ftext%3E%3C%2Fg%3E%3C%2Fg%3E%3C%2Fsvg%3E",
          "releaseDate": "2010.01.10",
          "releaseYear": 2010,
          "revenue": 0,
          "runtime": 0,
          "tagline": "tagline",
          "title": "Test title",
          "voteAvg": 1.5,
          "voteCount": 0,
        }
      }
    />
    <NavPanel
      direction="left"
      links={
        Array [
          Object {
            "active": true,
            "param": "genres",
            "title": "Genre: genre1, genre2",
          },
        ]
      }
      navBtn={null}
      primaryBrand={null}
      secondaryBrand="Films by"
    />
    <MoviesList
      movies={Array []}
      openDetailHandler={[Function]}
    />
  </React.Fragment>
</React.Fragment>
`;
